### ISSUE TITLE
Web Dashboard

### ISSUE THREAD

**Issue by Frederikam** [Frederikam/FredBoat#23]
I plan on creating a web dashboard primarily for fiddling around with bot permissions in a similar manner to the Discord permission system:
![image](https://cloud.githubusercontent.com/assets/2582617/18833678/31b987c0-83f2-11e6-98bb-e4ca12aee052.png)

The dashboard should have these features:

1. Standard Discord permissions for the server, text channels and voice channels
2. Additional bot permissions (play, stop, skip, NSFW, send commands, etc)
3. Manage permission overrides
4. OAuth2 login panel for authenticating with Discord
5. Maybe a Patreon link system
6. A simple list of servers

Further commands may follow, like kick ban and mute commands

My hope is that this major feature can be completed with the help of someone with experience on interactive websites. I will be able to provide an API to interface with, so the main concern is the front-end. Comment on this issue if you are interested.

**Comment by alperb** [comment]
I'm experienced about front end design and development. I can provide.

**Comment by Frederikam** [comment]
That'd be great! The site is coded with React.js. Got any experience with that? Otherwise, it's not too difficult to learn.

The repo for the web project is located at https://github.com/Frederikam/FredBoat-Web

Master is automatically built to https://fredboat.com/

**Comment by alperb** [comment]
I don't have any experience with React.

**Comment by Frederikam** [comment]
React is fairly easy to learn. Consider the fact that this is my first
non-static website.

https://facebook.github.io/react/

**Comment by celanthe** [comment]
@Frederikam If you're still in need of this, I know a bit of React and UI/UX design. I would love to help out with this.

**Comment by Frederikam** [comment]
Hey there @celanthe. Yes, this is still on the table and I would love some help, though I'm still wondering about how I am going to build the API gateway.

Be sure to ask if you need more info.

**Comment by celanthe** [comment]
@Frederikam I am wondering if https://github.com/celanthe/ReactJS-AdminLTE would be a good jumping off place to start for said dashboard? 

Unsure as to the gateway!

**Comment by Frederikam** [comment]
The API gateway is backend territory so it's probably something I'll end up dealing with. It's just impossible to finish this issue without it.

That React based template looks really neat and could be a useful starting point if you choose to delve into that. We don't need to really worry about having user customizable widgets and things like that, though.

**Comment by celanthe** [comment]
@Frederikam Fantastic! I'll delve into this over the next couple of weeks.

**Comment by celanthe** [comment]
@Frederikam slow progress, but still some progress. ^_^ Curious as to how you want the server permissions handled? For the sake of missing the back-end piece, would the commands from Discord need to actually do anything at this point?

I am assuming that the permissions in question are those listed at: https://discordapp.com/developers/docs/topics/permissions#role-object

EG: If someone clicks on the 'Discord Permissions' page in the sidebar, from there would it be enough to list the permissions for now in a table/with icons or somesuch until the back end is figured out?

![screenshot 2017-03-10 09 04 32](https://cloud.githubusercontent.com/assets/11055235/23768520/3470e3d6-0571-11e7-972c-f690ff6dba52.png)

**Comment by Frederikam** [comment]
Looks like you're off to a good start!

Here's some quick feedback:

1. We don't need to tell users that they are online. This is not a social platform.
2. I don't plan for this dashboard to be able to handle Discord permissions. That would be tricky to implement and would require maintenance for each new Discord permission.
3. Overrides are quite low priority at the moment, but if we do wind up getting them, we'll want to use the same layout as Discord rather than having a separate tab.
4. Sync with Discord? Not sure what that is supposed to be. Do you mean account authentication? People will need to authenticate to even view the dashboard.
5. The user must have some way of switching the server they are looking at managing.

I'm not sure what you are asking in regards to how we handle role permissions. In essence, we are just cloning Discord's UI for permissions, only with our own set of permissions. I'll need to think about exactly which permissions we want.

The base URL of the gateway will be `https://fredboat.com/api/`

I currently only intend to use the `identity` and `identity.guilds` scopes for oauth. The latter allows us to retrieve a list of guilds the user is in, and which permissions they have there.

**Comment by celanthe** [comment]
@Frederikam Ohh I see. I was making this much harder than it has to be, for sure! Wonderful. I'll get back to this over the weekend, for sure. :) That makes this a lot simpler.

**Comment by celanthe** [comment]
@Frederikam just wanted to update--should have some time to work on this next weekend for a bit. I work f/t so I'll progress through as time allows! Look forward to working on it.

**Comment by Frederikam** [comment]
Sounds good! There's no need to rush it though. I got my studies (and
currently also exams!) to worry about, so getting around to doing the
gateway will take some time.

**Comment by celanthe** [comment]
@Frederikam here's two different versions of what I've got so far. Also changed the user section to get rid of the social platform features, but that's apparently somehow borked the header. I'll figure that out and fix it for sure. 

![screenshot 2017-03-17 09 37 26](https://cloud.githubusercontent.com/assets/11055235/24028317/ca29efa0-0b32-11e7-9af3-e389703799d1.png)
![screenshot 2017-03-17 09 37 37](https://cloud.githubusercontent.com/assets/11055235/24028315/ca28d5ac-0b32-11e7-9d67-fd3b6ae3a13c.png)
![screenshot 2017-03-17 10 13 49](https://cloud.githubusercontent.com/assets/11055235/24028316/ca29c48a-0b32-11e7-9699-e0cb415d5662.png)
![screenshot 2017-03-17 10 13 57](https://cloud.githubusercontent.com/assets/11055235/24028319/ca2a71dc-0b32-11e7-9faa-bbb5cc49fc5b.png)




Not sure if either of these attached approaches is what is intended here. If I'm entirely off base, just let me know and I'll re-work this to be more appropriate. I'm still kind of new to programming, so I'm aware there's something also buggy in my code that's not giving me two side-by-side columns. I'll need to fix that. 

Anyway. Here's some pictures. Example A has the aforementioned checkboxes, but that might lead people to think they can actually manage permissions through the dash, which is something we are avoiding. If going for the second approach [with the panels linking out to OAuth2 perhaps?? in order to open the discord App/the bot and its applicable discord permissions] that would probably be easier. 

Again if I've completely missed the mark I apologize. ^^;;

**Comment by Frederikam** [comment]
There are quite a few problems here:

- [ ] The avatar is too big and there's no form of padding.
- [ ] You're mixing the guild selection with the guild settings. It's really odd to have it there, seeing from a UX design perspective.
- [ ] You're not using react-router for this page. The route should be something like `/guild/:guildId/manage`.
- [ ] Why is the home tab on the bottom? Most designers would put it at the top of the list.
- [ ] The home page is a mess. Why have those big unpadded banners when you can just style some simple buttons? We have an about page btw.
- [ ] Please use the same checkboxes that we use on the master branch. They are identical to what Discord uses.

It may be simpler to simply just fork master and continue my work. We already have react-router and checkmark components in place

**Comment by celanthe** [comment]
@Frederikam Thanks for the feedback. I'll revisit this at a later date and follow your advice for forking the master repo.

**Comment by averagenamespace** [comment]
So, because ASP.NET Core, I've made this:
![image](https://cloud.githubusercontent.com/assets/26074740/24070070/a9cd5dee-0c1a-11e7-9e2b-846170aff380.png)
![image](https://cloud.githubusercontent.com/assets/26074740/24070073/b32a2f8e-0c1a-11e7-831e-c3f244d15355.png)
It's a custom version of the Web Application template.
I'm almost done setting up Discord OAuth2 through ASP.NET Identities.

**Comment by Frederikam** [comment]
@averagenamespace looks really great! I particularly like the first image.

The system I had in mind is where the front-end is located on a CDN (namely netlify) and where API calls are made to `https://fredboat.com/api/some_route`.

Though a big problem with this is that it splits the stack into .NET and Java. Not completely sure how that is going to work. @Nanabell told me he would like to help if only he did Java.

**Comment by averagenamespace** [comment]
@Frederikam Has anything happened on the API end? I was assuming the API was going to be hosted separately (Java/whatever API backend communicates with ASP/React frontend), but if nothing has happened I can easily add some API controllers to this project, for example:
![image](https://cloud.githubusercontent.com/assets/26074740/24076260/25f3a74a-0c91-11e7-983f-6ff1dfbd471e.png)

**Comment by averagenamespace** [comment]
I see an application stack like this:
![image](https://cloud.githubusercontent.com/assets/26074740/24076357/0c85476c-0c93-11e7-8fbe-d36b69437fe4.png)

**Comment by Frederikam** [comment]
The problem with this is that FredBoat will eventually be dockerized into many different shards. My plan was for the API server to also manage those.

However, I'm not sure how your second layer fits in. By the third layer do you mean the bot itself?

**Comment by averagenamespace** [comment]
The middle and top layer (.NET) are the same application, I was going to
provide a wrapper (middle layer) that the dashboard uses.

**Comment by averagenamespace** [comment]
For now, I'll focus on building the web dashboard. Do you have a general idea of what the URLs are for the API endpoints?
For the requirements you stated in the first post:
I don't get what you mean by the first one, do you mean like a role editor? I don't think I can do that with the Discord API.
I can sort out updating what the bot can and cannot do from the dashboard, I just need a URL I can call (POST preferred) with a data object containing the new permissions (that I would assume update your local DB).

**Comment by averagenamespace** [comment]
Basic logic is this for the startup:
OAuth2 authorization (using the FredBoat main bot account as the app?)
Using the authorization, I GET `/users/@me/guilds` to get an array of basic guild objects, which has the guild ID. Then I GET `/guilds/<the guild ID>` to find more detailed info, including an array of `guild member ` objects. I'll search that for a member that contains FredBoat's ID. If it does, I'll show the control dash, if not, I'll display a jumbotron saying some cheesy message from a list of messages.

**Comment by averagenamespace** [comment]
@Frederikam Look what I just found 👀 https://github.com/RogueException/Discord.OAuth2

**Comment by averagenamespace** [comment]
Progress:
![image](https://cloud.githubusercontent.com/assets/26074740/24078729/9e5b1922-0cda-11e7-9b29-d77874bdeabb.png)
![image](https://cloud.githubusercontent.com/assets/26074740/24078746/f10f0660-0cda-11e7-84a9-315309b06f99.png)
![image](https://cloud.githubusercontent.com/assets/26074740/24078766/6b214166-0cdb-11e7-844c-0ae2af1b354d.png)
![image](https://cloud.githubusercontent.com/assets/26074740/24078769/946f0594-0cdb-11e7-8bf0-24b798d23614.png)

**Comment by Frederikam** [comment]
It is not desirable to add password/email auth. This makes the dashboard
needlessly complicated. It would mean that we would have to ensure that
passwords are safely hashed when stored, etc. Not to mention that we would
need to re-authenticate a user if they choose to revoke access:
[image: Inline images 1]

A user should *only* have access to the dashboard if they have a valid
bearer token stored in their browser.

Additionally, I don't feel like we are on the same page in regards to the
stack.
[image: Inline images 2]
I plan to scale FredBoat with docker containers. At some point, we'll need
a controller for these that can manage the shards (check if working
properly, rolling restarts, etc). My initial thought was to make this the
same program as the one serving API requests, but I don't see how that fits
in with your .NET program. Ideas?

**Comment by averagenamespace** [comment]
Yeah, I see what you're saying. I'll change it so that they are identified
by OAuth2 bearer tokens.

If you can provide endpoints, I can roll a socket manager into the middle
layer?

**Comment by Frederikam** [comment]
Socket manager? What do you mean by that?

**Comment by averagenamespace** [comment]
Failed at phone typing, I meant docker.

**Comment by Frederikam** [comment]
Sorry, but I fail to realize how that would work. API endpoints to control the swarm?

**Comment by averagenamespace** [comment]
Endpoints for each docker so that I can ping it and receive an
acknowledgement along with status details?

**Comment by Frederikam** [comment]
I could add endpoints on the shards, but how would you know which one of them to contact? Say you needed shard 23? Should I add an endpoint for that in the swarm controller?

**Comment by averagenamespace** [comment]
@Frederikam Exactly. Example:
`https://fredboat.com/api/shards/23`?

**Comment by averagenamespace** [comment]
I need some constants as well like:
* How often to ping each shard
* What a successful ping would look like
* What to do in case a shard doesn't return a successful ping (endpoints for restart etc)

**Comment by Frederikam** [comment]
A shard would return one of the following statuses:
https://github.com/DV8FromTheWorld/JDA/blob/master/src/main/java/net/dv8tion/jda/core/JDA.java#L40-L70

Probably in a format like this:
```json
{
    "shardId": 23,
    "status": "CONNECTED",
    "guildCount": 1008,
    "playersPlaying": 70,
    "baseUrl": "http://123:123:123:123:3523"
}
```

The endpoint would be `/api/shard/:shardId`. (note: shard is singular)

A dead shard is usually not playing anything or has no guilds. The status could be something like `ATTEMPTING_TO_RECONNECT` or `LOADING_SUBSYSTEMS`, but the shards *should* usually recover from this after a short while, but is sometimes not the case. 

I'm not sure how a shard would be restarted. I'm no docker expert.

**Comment by averagenamespace** [comment]
And if the shard sends no reply, is that a critical error (think red text), because the API should return data even if the bot is down? (This is for the owner dash)
`

**Comment by Frederikam** [comment]
If the shard doesn't respond it has likely crashed the JVM, or experiencing network issues. That should be considered critical.

**Comment by averagenamespace** [comment]
I don't know how to manage docker containers, do you mean the FredBoat (the Java API) API will manage the docker containers and I can call an endpoint with relevant auth to restart a dshard (Docker shard)?

**Comment by Frederikam** [comment]
I can probably take care of the docker swarm management, you'd just need to
make sure you are getting info from the correct shards.

**Comment by averagenamespace** [comment]
20/03/2017 Update:
* Removed ID by email, now identifying by Discord OAuth2 App User ID
* Removed all views under Manage and removed the ManageController
* Work started on dashboard, trying to implement AJAX calls to update the display (thinking of using React.js for the dashboard connecting to an ASP.NET backend)

**Comment by averagenamespace** [comment]
For the meantime; is there any (any at all), any endpoints I can call right now to check if FredBoat is online?

**Comment by Frederikam** [comment]
A status endpoint is currently available from http://api.fredboat.com:1356/stats

I have made some changes to my scaling strategy. Instead of creating one huge swarm I'm now thinking of putting up buffering sender nodes around the globe, which would lessen the load on my machines in Paris while decreasing packet loss.

**Comment by Alkaris** [comment]
This web dashboard do you HAVE to download and host the Web API or is there a publicly available Web Dashboard that you can use to sign-in to to configure FredBoat? because it sounds really tedious to just download and host your own accessible web host service just to use and customize this bot.

**Comment by Frederikam** [comment]
We haven't really decided on a solution for this problem yet. Perhaps it
would be ideal to leave an option on the dashboard to allow third party
bots to read the settings from a guild, but this can not really be
accomplished with a direct database connection like I would prefer.

**Comment by Frederikam** [comment]
Though you *will* need to set up an SQL database regardless.

**Comment by nobobo1234** [comment]
A little suggestion for this.

- Add smth like a dj role to the dashboard, so people can easily select a role from a dropdown menu
- Add a thing that people can select a music channel (with a dropdown menu) where the bot is only able to talk in
- Would like something to see like the dyno dashboard`
[Imgur link](https://i.imgur.com/VUZL8Ss.png) like this
Also I would like (like dyno's dashboard) a dashboard that is avaible for everyone

**Comment by Frederikam** [comment]
I was thinking of something more flexible though, like the Discord permission system

**Comment by nobobo1234** [comment]
Okay. But i would be cool for the music bot itself, with or without web dashboard. Especially for the dj role. I would really love if you would add some sort of command to that

**Comment by Frederikam** [comment]
Though we may want to simplify permissions, as many admins are already
having trouble managing Discord permissions.

**Comment by nobobo1234** [comment]
Okay but would a command be nice, automatically adding a dj role and add dj's to it by saying ?dj add <name> or ?adddj <name>. That isn't hard for people to understand right?

**Comment by Frederikam** [comment]
FredBoat is not going to automatically add roles to people's servers,
that's too intrusive.

**Comment by nobobo1234** [comment]
Really? Many bots written in js and python have that. But if the user can create a role itself or through some other form of moderation bot and than add djs by doing some command? Or is that too hard?

**Comment by Frederikam** [comment]
Maybe FredBoat could implicitly allow anyone with the "DJ" role, but I am going to have to trust admins to assign and create that role.

**Comment by nobobo1234** [comment]
Is that possible right now? To only let people with a specific role talk to FredBoat?

**Comment by Alkaris** [comment]
I saw someone speak of a similar thing that a bot can do automatic role selection on users for Patreon contributors when they donate and announce what they donated and when they donated using timestamps.

**Comment by Frederikam** [comment]
Yes, the Patreon API has webhook support, but in which way is that related to the dashboard?
